agent/subscribe/subscribe.go

```diff
@@ -0,0 +1,212 @@
+type Server struct {
+	Backend Backend
+	Logger  Logger
+}
+
+func NewServer(backend Backend, logger Logger) *Server {
+	return &Server{Backend: backend, Logger: logger}
+}
+
+type Logger interface {
+	Trace(msg string, args ...interface{})
+	With(args ...interface{}) hclog.Logger
+}
+
+var _ pbsubscribe.StateChangeSubscriptionServer = (*Server)(nil)
+
+type Backend interface {
+	ResolveToken(token string) (acl.Authorizer, error)
+	Forward(dc string, f func(*grpc.ClientConn) error) (handled bool, err error)
+	Subscribe(req *stream.SubscribeRequest) (*stream.Subscription, error)
+}
+
+func (h *Server) Subscribe(req *pbsubscribe.SubscribeRequest, serverStream pbsubscribe.StateChangeSubscription_SubscribeServer) error {
+	logger := h.newLoggerForRequest(req)
+	handled, err := h.Backend.Forward(req.Datacenter, forwardToDC(req, serverStream, logger))
+	if handled || err != nil {
+		return err
+	}
+
+	logger.Trace("new subscription")
+	defer logger.Trace("subscription closed")
+
+	// Resolve the token and create the ACL filter.
+	// TODO: handle token expiry gracefully...
+	authz, err := h.Backend.ResolveToken(req.Token)
+	if err != nil {
+		return err
+	}
+
+	sub, err := h.Backend.Subscribe(toStreamSubscribeRequest(req))
+	if err != nil {
+		return err
+	}
+	defer sub.Unsubscribe()
+
+	ctx := serverStream.Context()
+
+	elog := &eventLogger{logger: logger}
+	for {
+		events, err := sub.Next(ctx)
+		switch {
+		case errors.Is(err, stream.ErrSubscriptionClosed):
+			logger.Trace("subscription reset by server")
+			return status.Error(codes.Aborted, err.Error())
+		case err != nil:
+			return err
+		}
+
+		events = filterStreamEvents(authz, events)
+		if len(events) == 0 {
+			continue
+		}
+
+		elog.Trace(events)
+		e := newEventFromStreamEvents(req, events)
+		if err := serverStream.Send(e); err != nil {
+			return err
+		}
+	}
+}
+
+// TODO: can be replaced by mog conversion
+func toStreamSubscribeRequest(req *pbsubscribe.SubscribeRequest) *stream.SubscribeRequest {
+	return &stream.SubscribeRequest{
+		Topic: req.Topic,
+		Key:   req.Key,
+		Token: req.Token,
+		Index: req.Index,
+	}
+}
+
+func forwardToDC(
+	req *pbsubscribe.SubscribeRequest,
+	serverStream pbsubscribe.StateChangeSubscription_SubscribeServer,
+	logger Logger,
+) func(conn *grpc.ClientConn) error {
+	return func(conn *grpc.ClientConn) error {
+		logger.Trace("forwarding to another DC")
+		defer logger.Trace("forwarded stream closed")
+
+		client := pbsubscribe.NewStateChangeSubscriptionClient(conn)
+		streamHandle, err := client.Subscribe(serverStream.Context(), req)
+		if err != nil {
+			return err
+		}
+
+		for {
+			event, err := streamHandle.Recv()
+			if err != nil {
+				return err
+			}
+			if err := serverStream.Send(event); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+// filterStreamEvents to only those allowed by the acl token.
+func filterStreamEvents(authz acl.Authorizer, events []stream.Event) []stream.Event {
+	// TODO: when is authz nil?
```

Comment:
When ACLs are disabled IIRC? It could just be defensive but we certainly have places in Endpoint handlers where authz resolves to nil if ACLs are disabled.

Reply:
aha, yes, that is the case. I guess I scanned over that code too quickly. I'll update this to a comment which says so